File: naming_examples.py
"""
Examples of Improving Variable and Function Names
This file demonstrates how to transform poorly named variables and functions
into clear, descriptive names that improve code readability.
"""

# ==========================================
# Example 1: Single Letter Variables
# ==========================================

# ❌ BAD: Single letter variables
def calc(a, b, c):
    t = a * b
    r = t * (1 + c)
    return r

# ✅ GOOD: Descriptive variable names
def calculate_total_price_with_tax(price, quantity, tax_rate):
    subtotal = price * quantity
    total_with_tax = subtotal * (1 + tax_rate)
    return total_with_tax


# ==========================================
# Example 2: Vague Names
# ==========================================

# ❌ BAD: Vague, unclear names
def process(data):
    result = []
    for item in data:
        if item > 0:
            result.append(item * 2)
    return result

# ✅ GOOD: Clear, specific names
def double_positive_numbers(numbers):
    doubled_positive_numbers = []
    for number in numbers:
        if number > 0:
            doubled_positive_numbers.append(number * 2)
    return doubled_positive_numbers


# ==========================================
# Example 3: Abbreviations
# ==========================================

# ❌ BAD: Unclear abbreviations
def get_usr_info(usr_id):
    usr_data = {"id": usr_id, "nm": "John", "addr": "123 Main St"}
    return usr_data

# ✅ GOOD: Full, descriptive words
def get_user_information(user_id):
    user_data = {"id": user_id, "name": "John", "address": "123 Main St"}
    return user_data


# ==========================================
# Example 4: Function Names
# ==========================================

# ❌ BAD: Vague function name
def check(email):
    return "@" in email and "." in email

# ✅ GOOD: Function name describes what it does
def is_valid_email_format(email):
    return "@" in email and "." in email


# ❌ BAD: Unclear what function returns
def get(user_id):
    # ... database logic
    return True

# ✅ GOOD: Clear return value indication
def is_user_active(user_id):
    # ... database logic
    return True


# ==========================================
# Example 5: Loop Variables
# ==========================================

# ❌ BAD: Generic loop variables
def sum_prices(items):
    s = 0
    for i in items:
        s += i["p"]
    return s

# ✅ GOOD: Descriptive loop variables
def calculate_total_price(products):
    total_price = 0
    for product in products:
        total_price += product["price"]
    return total_price


# ==========================================
# Example 6: Boolean Variables
# ==========================================

# ❌ BAD: Unclear boolean names
def check_user(user):
    flag = user.get("age") >= 18
    status = user.get("verified")
    return flag and status

# ✅ GOOD: Clear boolean names with is/has/can prefix
def can_user_access_content(user):
    is_adult = user.get("age") >= 18
    is_verified = user.get("verified")
    return is_adult and is_verified


# ==========================================
# Example 7: Constants
# ==========================================

# ❌ BAD: Unclear constant purpose
MAX = 100
MIN = 0
RATE = 0.15

# ✅ GOOD: Clear constant purpose
MAX_LOGIN_ATTEMPTS = 100
MIN_PASSWORD_LENGTH = 0
DEFAULT_TAX_RATE = 0.15


# ==========================================
# Example 8: Class and Method Names
# ==========================================

# ❌ BAD: Vague class and method names
class Manager:
    def do(self, data):
        return data.upper()
    
    def handle(self, items):
        return sorted(items)

# ✅ GOOD: Descriptive class and method names
class UserAccountManager:
    def convert_username_to_uppercase(self, username):
        return username.upper()
    
    def sort_users_by_creation_date(self, users):
        return sorted(users)


# ==========================================
# Example 9: Temporary Variables
# ==========================================

# ❌ BAD: Generic temp variables
def process_order(order):
    temp1 = order["items"]
    temp2 = sum(item["price"] for item in temp1)
    temp3 = temp2 * 1.1
    return temp3

# ✅ GOOD: Meaningful temporary variables
def calculate_order_total_with_service_fee(order):
    order_items = order["items"]
    items_subtotal = sum(item["price"] for item in order_items)
    total_with_service_fee = items_subtotal * 1.1
    return total_with_service_fee


# ==========================================
# Example 10: Context-Specific Names
# ==========================================

# ❌ BAD: Names without context
def update(id, val):
    # What are we updating?
    database.update(id, val)

# ✅ GOOD: Context-specific names
def update_user_email_address(user_id, new_email_address):
    # Clear what we're updating
    database.update(user_id, new_email_address)


# ==========================================
# Example 11: Collection Names
# ==========================================

# ❌ BAD: Generic collection names
def filter_data(list1):
    list2 = []
    for x in list1:
        if x.active:
            list2.append(x)
    return list2

# ✅ GOOD: Descriptive collection names
def filter_active_users(all_users):
    active_users = []
    for user in all_users:
        if user.active:
            active_users.append(user)
    return active_users


# ==========================================
# Example 12: Time-Related Variables
# ==========================================

# ❌ BAD: Unclear time units
def wait(t):
    import time
    time.sleep(t)

# ✅ GOOD: Clear time units in name
def wait_for_seconds(duration_in_seconds):
    import time
    time.sleep(duration_in_seconds)


if __name__ == "__main__":
    print("✨ Variable and Function Naming Examples")
    print("=" * 50)
    
    # Test Example 1
    print("\nExample 1: Calculate Price with Tax")
    total = calculate_total_price_with_tax(100, 2, 0.08)
    print(f"Total: ${total:.2f}")
    
    # Test Example 2
    print("\nExample 2: Double Positive Numbers")
    numbers = [-5, 10, -3, 20, 15]
    result = double_positive_numbers(numbers)
    print(f"Doubled positive numbers: {result}")
    
    # Test Example 4
    print("\nExample 4: Email Validation")
    email = "test@example.com"
    is_valid = is_valid_email_format(email)
    print(f"Is '{email}' valid? {is_valid}")
    
    print("\n" + "=" * 50)
    print("✅ All examples demonstrate improved naming!")

File: naming_examples.js
/**
 * JavaScript Examples of Improving Variable and Function Names
 * This file demonstrates how to transform poorly named variables and functions
 * into clear, descriptive names that improve code readability.
 */

// ==========================================
// Example 1: Single Letter Variables
// ==========================================

// ❌ BAD: Single letter variables
function calc(a, b, c) {
    const t = a * b;
    const r = t * (1 + c);
    return r;
}

// ✅ GOOD: Descriptive variable names
function calculateTotalPriceWithTax(price, quantity, taxRate) {
    const subtotal = price * quantity;
    const totalWithTax = subtotal * (1 + taxRate);
    return totalWithTax;
}


// ==========================================
// Example 2: Array Methods
// ==========================================

// ❌ BAD: Vague arrow function parameters
const filterData = (arr) => arr.filter(x => x > 10);

// ✅ GOOD: Clear parameter names
const filterNumbersGreaterThanTen = (numbers) => 
    numbers.filter(number => number > 10);


// ==========================================
// Example 3: Object Properties
// ==========================================

// ❌ BAD: Unclear abbreviations
const usr = {
    id: 1,
    nm: "John",
    em: "john@example.com",
    dob: "1990-01-01"
};

// ✅ GOOD: Full, descriptive property names
const userProfile = {
    userId: 1,
    fullName: "John",
    emailAddress: "john@example.com",
    dateOfBirth: "1990-01-01"
};


// ==========================================
// Example 4: Event Handlers
// ==========================================

// ❌ BAD: Generic handler names
function handle(e) {
    e.preventDefault();
    // Do something
}

// ✅ GOOD: Descriptive handler names
function handleFormSubmission(event) {
    event.preventDefault();
    // Submit form data
}

function handleUserLoginClick(event) {
    event.preventDefault();
    // Process login
}


// ==========================================
// Example 5: Async Functions
// ==========================================

// ❌ BAD: Unclear async function purpose
async function get(id) {
    const response = await fetch(`/api/${id}`);
    return response.json();
}

// ✅ GOOD: Clear async function purpose
async function fetchUserDataById(userId) {
    const response = await fetch(`/api/users/${userId}`);
    return response.json();
}


// ==========================================
// Example 6: Boolean Variables
// ==========================================

// ❌ BAD: Unclear boolean names
function checkAuth(user) {
    const flag = user.loggedIn;
    const status = user.verified;
    return flag && status;
}

// ✅ GOOD: Clear boolean names with is/has/can prefix
function canUserAccessProtectedContent(user) {
    const isLoggedIn = user.loggedIn;
    const isVerified = user.verified;
    return isLoggedIn && isVerified;
}


// ==========================================
// Example 7: Constants
// ==========================================

// ❌ BAD: Unclear constant purpose
const MAX = 5;
const LIMIT = 100;
const TIME = 3000;

// ✅ GOOD: Clear constant purpose
const MAX_RETRY_ATTEMPTS = 5;
const API_REQUEST_LIMIT = 100;
const NOTIFICATION_TIMEOUT_MS = 3000;


// ==========================================
// Example 8: Class Names
// ==========================================

// ❌ BAD: Vague class names
class Manager {
    constructor(data) {
        this.data = data;
    }
    
    do() {
        return this.data.map(x => x * 2);
    }
}

// ✅ GOOD: Descriptive class and method names
class ShoppingCartManager {
    constructor(cartItems) {
        this.cartItems = cartItems;
    }
    
    calculateItemsWithDoubledQuantity() {
        return this.cartItems.map(item => item * 2);
    }
}


// ==========================================
// Example 9: Promise Chains
// ==========================================

// ❌ BAD: Generic variable names in promise chains
function badPromiseChainExample() {
    // In a real app: 
    // fetch('/api/data')
    //     .then(res => res.json())
    //     .then(d => {
    //         const r = d.filter(x => x.active);
    //         return r;
    //     });
}

// ✅ GOOD: Descriptive names in promise chains
function goodPromiseChainExample() {
    // In a real app:
    // fetch('/api/users')
    //     .then(response => response.json())
    //     .then(allUsers => {
    //         const activeUsers = allUsers.filter(user => user.active);
    //         return activeUsers;
    //     });
}


// ==========================================
// Example 10: Destructuring
// ==========================================

// ❌ BAD: Destructuring with unclear names
function process({ a, b, c }) {
    return a + b + c;
}

// ✅ GOOD: Destructuring with clear names
function calculateOrderTotal({ basePrice, shippingCost, taxAmount }) {
    return basePrice + shippingCost + taxAmount;
}


// ==========================================
// Example 11: Callback Functions
// ==========================================

// ❌ BAD: Generic callback names
function doSomething(data, cb) {
    const result = data.map(x => x * 2);
    cb(result);
}

// ✅ GOOD: Descriptive callback parameter names
function processUserData(users, onProcessingComplete) {
    const processedUsers = users.map(user => user * 2);
    onProcessingComplete(processedUsers);
}


// ==========================================
// Example 12: Reducer Functions
// ==========================================

// ❌ BAD: Generic reducer names
function badReducerExample(items) {
    const result = items.reduce((acc, curr) => {
        return acc + curr.val;
    }, 0);
    return result;
}

// ✅ GOOD: Clear reducer variable names
function calculateTotalPriceFromItems(orderItems) {
    const totalPrice = orderItems.reduce((runningTotal, currentItem) => {
        return runningTotal + currentItem.price;
    }, 0);
    return totalPrice;
}


// ==========================================
// Example 13: State Management
// ==========================================

// ❌ BAD: Unclear state variable names
function badStateExample() {
    let state = false;
    let data = null;
    let flag = true;
    return { state, data, flag };
}

// ✅ GOOD: Descriptive state variable names
function goodStateExample() {
    let isUserAuthenticated = false;
    let loadedUserProfile = null;
    let shouldShowNotification = true;
    return { isUserAuthenticated, loadedUserProfile, shouldShowNotification };
}


// ==========================================
// Example 14: Time and Intervals
// ==========================================

// ❌ BAD: Unclear timing variables
function badTimingExample() {
    const t = 5000;
    // Would call: setTimeout(() => {}, t);
}

// ✅ GOOD: Clear timing with units
function goodTimingExample() {
    const AUTO_SAVE_INTERVAL_MS = 5000;
    function saveUserProgress() {
        // Save user data
    }
    // Would call: setTimeout(() => { saveUserProgress(); }, AUTO_SAVE_INTERVAL_MS);
}


// ==========================================
// Example 15: DOM Manipulation
// ==========================================

// ❌ BAD: Generic element names
function badDOMExample() {
    if (typeof document !== 'undefined') {
        const el = document.querySelector('.item');
        const btn = document.getElementById('btn');
    }
}

// ✅ GOOD: Descriptive element names
function goodDOMExample() {
    if (typeof document !== 'undefined') {
        const productCardElement = document.querySelector('.product-card');
        const submitOrderButton = document.getElementById('submit-order-btn');
    }
}


// Test the examples
if (typeof window === 'undefined') {
    // Node.js environment - run tests
    console.log('✨ JavaScript Variable and Function Naming Examples');
    console.log('='.repeat(50));
    
    // Test Example 1
    console.log('\nExample 1: Calculate Price with Tax');
    const total = calculateTotalPriceWithTax(100, 2, 0.08);
    console.log(`Total: $${total.toFixed(2)}`);
    
    // Test Example 2
    console.log('\nExample 2: Filter Numbers > 10');
    const numbers = [5, 15, 8, 20, 12];
    const filtered = filterNumbersGreaterThanTen(numbers);
    console.log(`Filtered numbers: ${filtered}`);
    
    // Test Example 10
    console.log('\nExample 10: Calculate Order Total');
    const orderTotal = calculateOrderTotal({
        basePrice: 100,
        shippingCost: 15,
        taxAmount: 8.5
    });
    console.log(`Order total: $${orderTotal.toFixed(2)}`);
    
    console.log('\n' + '='.repeat(50));
    console.log('✅ All examples demonstrate improved naming!');
}
